File: scaperbuild/frontend/scripts/wire-solar-cms.mjs
#!/usr/bin/env node
/** Wire SolarPanelBirdProofingPage to Laravel CMS via dedicated solar components. */
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const filePath = path.join(__dirname, '../src/pages/services/SolarPanelBirdProofingPage.jsx')

let src = fs.readFileSync(filePath, 'utf8')
if (src.includes('usePageCms(')) {
  console.log('Solar page already wired')
  process.exit(0)
}

const imports = `import CmsHtml from '../../components/home/CmsHtml.jsx'
import SolarAdvantagesCards from '../../components/solar/SolarAdvantagesCards.jsx'
import SolarBlogPosts from '../../components/solar/SolarBlogPosts.jsx'
import SolarFaqSection from '../../components/solar/SolarFaqSection.jsx'
import SolarPanelCmsBinder from '../../components/solar/SolarPanelCmsBinder.jsx'
import SolarPanelCmsStyles from '../../components/solar/SolarPanelCmsStyles.jsx'
import SolarServiceCards from '../../components/solar/SolarServiceCards.jsx'
import SolarSignsCards from '../../components/solar/SolarSignsCards.jsx'
import SolarWhyChooseCards from '../../components/solar/SolarWhyChooseCards.jsx'
import SolarWhyEssentialCards from '../../components/solar/SolarWhyEssentialCards.jsx'
import usePageCms from '../../hooks/usePageCms.js'
import { cmsText } from '../../utils/cmsMedia.js'
`

src = src.replace(
  /import PhoneNumberInput from '\.\.\/\.\.\/components\/PhoneNumberInput\.jsx'\n/,
  `import PhoneNumberInput from '../../components/PhoneNumberInput.jsx'\n${imports}`,
)

src = src.replace(
  /export default function SolarPanelBirdProofingPage\(\) \{\n  usePageMeta\('solar_panel_bird_proofing'\)/,
  `export default function SolarPanelBirdProofingPage() {
  const { page, content: c } = usePageCms('solar-panel-bird-proofing')
  usePageMeta('solar_panel_bird_proofing', page)`,
)

src = src.replace(
  /return \(\n    <>\n/,
  `return (
    <>
      <SolarPanelCmsStyles content={c} />
      <SolarPanelCmsBinder content={c} />
`,
)

function replaceBlock(src, startId, endBeforeId, replacement) {
  const re = new RegExp(
    `<div className="elementor-element elementor-element-${startId}[\\s\\S]*?(?=<div className="elementor-element elementor-element-${endBeforeId})`,
  )
  return src.replace(re, `${replacement}\n            `)
}

src = replaceBlock(
  src,
  '5602f2b',
  '972e1d3',
  '<SolarServiceCards items={c?.services?.items ?? []} />',
)

src = replaceBlock(
  src,
  '8f02689',
  '1f6db9f',
  '<SolarWhyEssentialCards items={c?.why_essential?.items ?? []} />',
)

// why essential section ends before next top container - fix: end before a047a50 section
src = src.replace(
  /<SolarWhyEssentialCards items=\{c\?\.why_essential\?\.items \?\? \[\]\} \/>[\s\S]*?(?=<section className="elementor-section elementor-top-section elementor-element elementor-element-a047a50)/,
  '<SolarWhyEssentialCards items={c?.why_essential?.items ?? []} />\n        ',
)

src = replaceBlock(
  src,
  '3d668ca',
  '62fdcad',
  '<SolarSignsCards items={c?.signs?.items ?? []} />',
)

src = src.replace(
  /<SolarSignsCards items=\{c\?\.signs\?\.items \?\? \[\]\} \/>[\s\S]*?(?=<div className="elementor-element elementor-element-62fdcad)/,
  '<SolarSignsCards items={c?.signs?.items ?? []} />\n            ',
)

src = replaceBlock(
  src,
  '62fdcad',
  '1f6db9f',
  '<SolarAdvantagesCards items={c?.advantages?.items ?? []} />',
)

// advantages section - end before process section; find next unique id after advantages
src = src.replace(
  /<SolarAdvantagesCards items=\{c\?\.advantages\?\.items \?\? \[\]\} \/>[\s\S]*?(?=<div className="elementor-element elementor-element-[a-f0-9]+ e-flex e-con-boxed e-con e-parent[^>]*data-id="(?!62fdcad)[a-f0-9]+"[^>]*>\s*<div className="e-con-inner">\s*<div className="elementor-element elementor-element-[a-f0-9]+ elementor-widget elementor-widget-heading"[^>]*>\s*<div className="elementor-widget-container">\s*<h2 className="elementor-heading-title elementor-size-default">Solar Panel Bird Proofing and Cleaning Process)/,
  '<SolarAdvantagesCards items={c?.advantages?.items ?? []} />\n        ',
)

src = replaceBlock(
  src,
  '2583f30',
  'c190105',
  '<SolarWhyChooseCards items={c?.why_choose?.items ?? []} />',
)

src = src.replace(
  /<div className="elementor-element elementor-element-c190105 elementor-widget elementor-widget-elementskit-blog-posts"[\s\S]*?<\/div>\s*<\/div>\s*<\/div>\s*<\/div>\s*<\/section>/,
  '<SolarBlogPosts blog={c?.blog} />\n        </section>',
)

src = src.replace(
  /<div className="elementor-element elementor-element-5a3541e elementor-widget elementor-widget-elementskit-accordion"[\s\S]*?<\/div>\s*<\/div>\s*<\/div>\s*<\/div>\s*<\/div>\s*<\/section>/,
  '<SolarFaqSection faq={c?.faq} />\n              </div>\n            </div>\n          </div>\n        </section>',
)

// Hero
src = src.replace(
  /<span className="elementor-icon-list-text">Get Affordable Bird Removal Service<\/span>/,
  `<span className="elementor-icon-list-text">{cmsText(c?.hero?.breadcrumb, 'Get Affordable Bird Removal Service')}</span>`,
)
src = src.replace(
  /<h2 className="elementor-heading-title elementor-size-default">Solar Panel Bird Protection Services<\/h2>/,
  `<h2 className="elementor-heading-title elementor-size-default">{cmsText(c?.hero?.title, 'Solar Panel Bird Protection Services')}</h2>`,
)
src = src.replace(
  /<h1 className="elementor-heading-title elementor-size-default">Get Affordable Bird Removal Services in Melbourne<\/h1>/,
  `<h1 className="elementor-heading-title elementor-size-default">{cmsText(c?.hero?.heading, 'Get Affordable Bird Removal Services in Melbourne')}</h1>`,
)

src = src.replace(
  /(<div className="elementor-element elementor-element-7a605c4 elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)[\s\S]*?(<\/div>\s*<\/div>\s*<\/div>\s*<\/div>\s*<\/div>\s*<\/section>)/,
  '$1\n                  <CmsHtml html={c?.hero?.intro} />\n                $2',
)

src = src.replace(
  /<h2 className="elementor-heading-title elementor-size-default">Complete Solar Panel Bird Protection Services <\/h2>/,
  `<h2 className="elementor-heading-title elementor-size-default">{cmsText(c?.services?.title, 'Complete Solar Panel Bird Protection Services')}</h2>`,
)

src = src.replace(
  /(<div className="elementor-element elementor-element-3440d1b elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)\s*<p>Keep your solar panels[\s\S]*?<\/p>(\s*<\/div>)/,
  '$1\n                  <CmsHtml html={c?.services?.intro} />$2',
)

src = src.replace(
  /(<div className="elementor-element elementor-element-6be9f35 elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)\s*<p>Invest in long-term[\s\S]*?<\/p>(\s*<\/div>)/,
  '$1\n                  <CmsHtml html={c?.services?.footer} />$2',
)

src = src.replace(
  /<h2 className="elementor-heading-title elementor-size-default">Why Bird Pest Control in Melbourne is Essential for Protecting Your Property\?<\/h2>/,
  `<h2 className="elementor-heading-title elementor-size-default">{cmsText(c?.why_essential?.title, 'Why Bird Pest Control in Melbourne is Essential for Protecting Your Property?')}</h2>`,
)

src = src.replace(
  /(<div className="elementor-element elementor-element-879c572 elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)\s*<p>Birds on your roof[\s\S]*?<\/p>(\s*<\/div>)/,
  '$1\n                  <CmsHtml html={c?.why_essential?.intro} />$2',
)

src = src.replace(
  /<h2 className="elementor-heading-title elementor-size-default">Frequently Asked Questions<\/h2>/,
  `<h2 className="elementor-heading-title elementor-size-default">{cmsText(c?.faq?.title, 'Frequently Asked Questions')}</h2>`,
)

fs.writeFileSync(filePath, src)
console.log('Wired solar-panel-bird-proofing')

File: scaperbuild/frontend/scripts/wire-solar-cms.py
#!/usr/bin/env python3
"""Safely wire SolarPanelBirdProofingPage to Laravel CMS."""
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
FILE = ROOT / 'frontend/src/pages/services/SolarPanelBirdProofingPage.jsx'


def container_bounds(content: str, element_id: str, from_pos: int = 0) -> tuple[int, int]:
    needle = f'<div className="elementor-element elementor-element-{element_id}'
    start = content.find(needle, from_pos)
    if start == -1:
        raise SystemExit(f'Container {element_id} not found')
    depth = 0
    i = start
    n = len(content)
    while i < n:
        if content.startswith('<div', i):
            depth += 1
            i = content.find('>', i) + 1
        elif content.startswith('</div>', i):
            depth -= 1
            i += 6
            if depth == 0:
                return start, i
        else:
            i += 1
    raise SystemExit(f'Unbalanced div for {element_id}')


def replace_elementor_container(content: str, element_id: str, replacement: str, from_pos: int = 0) -> str:
    start, end = container_bounds(content, element_id, from_pos)
    return content[:start] + replacement + content[end:]


def replace_range(content: str, start_id: str, end_id: str, replacement: str) -> str:
    start, _ = container_bounds(content, start_id)
    _, end = container_bounds(content, end_id, start)
    return content[:start] + replacement + content[end:]


def main():
    src = FILE.read_text()
    if 'usePageCms(' in src:
        print('Already wired')
        return

    imports = """import CmsHtml from '../../components/home/CmsHtml.jsx'
import SolarAdvantagesCards from '../../components/solar/SolarAdvantagesCards.jsx'
import SolarBlogPosts from '../../components/solar/SolarBlogPosts.jsx'
import SolarFaqSection from '../../components/solar/SolarFaqSection.jsx'
import SolarPanelCmsBinder from '../../components/solar/SolarPanelCmsBinder.jsx'
import SolarPanelCmsStyles from '../../components/solar/SolarPanelCmsStyles.jsx'
import SolarServiceCards from '../../components/solar/SolarServiceCards.jsx'
import SolarSignsCards from '../../components/solar/SolarSignsCards.jsx'
import SolarWhyChooseCards from '../../components/solar/SolarWhyChooseCards.jsx'
import SolarWhyEssentialCards from '../../components/solar/SolarWhyEssentialCards.jsx'
import usePageCms from '../../hooks/usePageCms.js'
import { cmsText } from '../../utils/cmsMedia.js'
"""
    src = src.replace(
        "import PhoneNumberInput from '../../components/PhoneNumberInput.jsx'\n",
        "import PhoneNumberInput from '../../components/PhoneNumberInput.jsx'\n" + imports,
    )
    src = src.replace(
        "export default function SolarPanelBirdProofingPage() {\n  usePageMeta('solar_panel_bird_proofing')",
        "export default function SolarPanelBirdProofingPage() {\n  const { page, content: c } = usePageCms('solar-panel-bird-proofing')\n  usePageMeta('solar_panel_bird_proofing', page)",
    )
    src = src.replace(
        'return (\n    <>\n',
        'return (\n    <>\n      <SolarPanelCmsStyles content={c} />\n      <SolarPanelCmsBinder content={c} />\n',
    )

    # Hero text fields
    src = src.replace(
        '<span className="elementor-icon-list-text">Get Affordable Bird Removal Service</span>',
        "<span className=\"elementor-icon-list-text\">{cmsText(c?.hero?.breadcrumb, 'Get Affordable Bird Removal Service')}</span>",
    )
    src = src.replace(
        '<h2 className="elementor-heading-title elementor-size-default">Solar Panel Bird Protection Services</h2>',
        "<h2 className=\"elementor-heading-title elementor-size-default\">{cmsText(c?.hero?.title, 'Solar Panel Bird Protection Services')}</h2>",
    )
    src = src.replace(
        '<h1 className="elementor-heading-title elementor-size-default">Get Affordable Bird Removal Services in Melbourne</h1>',
        "<h1 className=\"elementor-heading-title elementor-size-default\">{cmsText(c?.hero?.heading, 'Get Affordable Bird Removal Services in Melbourne')}</h1>",
    )
    src = re.sub(
        r'(<div className="elementor-element elementor-element-7a605c4 elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)\s*<p>Expert bird pest control[\s\S]*?</p>\s*(</div>)',
        r'\1\n                  <CmsHtml html={c?.hero?.intro} />\n                \2',
        src,
        count=1,
    )

    src = src.replace(
        '<h2 className="elementor-heading-title elementor-size-default">Complete Solar Panel Bird Protection Services </h2>',
        "<h2 className=\"elementor-heading-title elementor-size-default\">{cmsText(c?.services?.title, 'Complete Solar Panel Bird Protection Services')}</h2>",
    )
    src = re.sub(
        r'(<div className="elementor-element elementor-element-3440d1b elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)\s*<p>Keep your solar panels[\s\S]*?</p>\s*(</div>)',
        r'\1\n                  <CmsHtml html={c?.services?.intro} />\n                \2',
        src,
        count=1,
    )
    src = re.sub(
        r'(<div className="elementor-element elementor-element-6be9f35 elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)\s*<p>Invest in long-term[\s\S]*?</p>\s*(</div>)',
        r'\1\n                  <CmsHtml html={c?.services?.footer} />\n                \2',
        src,
        count=1,
    )
    src = src.replace(
        '<h2 className="elementor-heading-title elementor-size-default">Why Bird Pest Control in Melbourne is Essential for Protecting Your Property?</h2>',
        "<h2 className=\"elementor-heading-title elementor-size-default\">{cmsText(c?.why_essential?.title, 'Why Bird Pest Control in Melbourne is Essential for Protecting Your Property?')}</h2>",
    )
    src = re.sub(
        r'(<div className="elementor-element elementor-element-879c572 elementor-widget elementor-widget-text-editor"[^>]*>\s*<div className="elementor-widget-container">)\s*<p>Birds on your roof[\s\S]*?</p>\s*(</div>)',
        r'\1\n                  <CmsHtml html={c?.why_essential?.intro} />\n                \2',
        src,
        count=1,
    )
    src = src.replace(
        '<h2 className="elementor-heading-title elementor-size-default">Frequently Asked Questions</h2>',
        "<h2 className=\"elementor-heading-title elementor-size-default\">{cmsText(c?.faq?.title, 'Frequently Asked Questions')}</h2>",
    )

    # Card sections — replace from end to start so indices stay valid
    src = replace_range(src, '62fdcad', 'e86bf87', '<SolarAdvantagesCards items={c?.advantages?.items ?? []} />\n            ')
    src = replace_range(src, '3d668ca', 'cb13aae', '<SolarSignsCards items={c?.signs?.items ?? []} />\n            ')
    src = replace_range(src, '8f02689', '6b3e7c4', '<SolarWhyEssentialCards items={c?.why_essential?.items ?? []} />\n            ')
    src = replace_range(src, '5602f2b', '9f28e19', '<SolarServiceCards items={c?.services?.items ?? []} />\n            ')
    src = replace_elementor_container(src, '2583f30', '<SolarWhyChooseCards items={c?.why_choose?.items ?? []} />\n            ')
    src = replace_elementor_container(src, 'c190105', '<SolarBlogPosts blog={c?.blog} />\n                ')
    src = replace_elementor_container(src, '5a3541e', '<SolarFaqSection faq={c?.faq} />\n                ')

    FILE.write_text(src)
    print('Wired solar-panel-bird-proofing')


if __name__ == '__main__':
    main()
